README.md
# LDAP Query
LDAP-Query is a service which provides a REST API gateway to make queries against an LDAP directory.  It provides a read-only interface; changes to the directory are not supported.

It has specifically been tested against Active Directory, but there shouldn't be anything which is specific to AD so there is no obvious reason why it wouldn't work against other LDAP directories.

## Usage
Configuration is passed via flags.  If you need to query multiple domains, just run multiple instances of the application; remember to change the port that the application binds to.

| Flag              | Description                                                                                                  | Default Value |
| ----------------- | ------------------------------------------------------------------------------------------------------------ | ------------- |
| port              | Port the application listens on                                                                              | 9999          |
| allowed_sources   | Comma separated list of IPs to accept requests from                                                          | none          |
| directory_hosts   | Comma separated list of LDAP hosts to query; these should all be in the same domain                          | 9280          |
| directory_bind_dn | Full distinguished name of user account used to bind to the directory                                        | none          |
| directory_bind_pw | Password for the user account used to bind to the directory.  This does NOT need to be a privileged account. | none          |
| version           | Display application version information                                                                      | false         |
| service           | Manage Windows services; install, uninstall, start, and stop                                                 | none          |
| help              | Display help                                                                                                 | false         |
| debug             | Enable debug logging                                                                                         | false         |

````
ldap-queryd.exe --allowed_sources "172.16.124.34" --directory_hosts "192.168.1.22,192.168.1.56" --directory_bind_dn "CN=account1,CN=Users,DC=my,DC=domain" --directory_bind_pw "complex_password"
````

A machine with IP 172.16.124.34 is allowed to send request to LDAP hosts at 192.168.1.22 and 192.168.1.56 on the default port of 389, using the account details provided.

Once running you can run any query you want by sending a `POST` request to the `/search` endpoint with your query as the JSON payload.  Here is an example:

``` json
POST /search

{
    "filter": "(&(&(objectCategory=Person)(objectClass=User))(sn=skywalk*))",
    "scope": "sub",
    "base": "ou=xxx,dc=xxx,dc=xxx,dc=xx",
    "attributes": [
        "sAMAccountName",
        "cn",
        "givenName"
    ]
}
```

The `filter`, `base`, and `attributes` parameters are **required**.  The `scope` parameter is not required and will default to `base`.

No validation is carried out on the filter or attribute names, so if you don't get the results you expect make sure you check that they are correct.

:warning: If the object you are searching for has brackets in the name, either `(` or `)`, you will need to escape the filter.  So a filter like `(&(cn=my group (admins),dc=xxx,dc=xxx,dc=xxx)(objectCategory=group))` needs to be like this -> `(&(cn=my group \\28admins\\29,dc=xxx,dc=xxx,dc=xxx)(objectCategory=group))`.

To display the application version run the application with the `--version` flag.

### Metrics
Application metrics are exported in [Prometheus](https://prometheus.io/) format to the `/metrics` endpoint.

### Logs
#### Windows
When running as a service logs are sent to the `Application` Event Log with a `Source` of `LDAP-Query`.

When running interactively, logs will be sent to `Stdout`.

#### Linux/Docker
Logs will be sent to `Stdout`.

## Running in production
It is best to setup `ldap-query` to run as a service/daemon.

### Windows
You can install the service like so:

```
ldap-queryd.exe --service install
```

Uninstall like so:

```
ldap-queryd.exe --service uninstall
```

You'll probably want to ensure that the service is set to restart on failures, so check those settings.

When running as a service, logs will be sent to the `Application` Event Log using a `Source` of `LDAP-Query`.

When running interactively, logs will be sent to `Stdout`.

### Linux
Not sure, but likely just Docker?

**TODO** - Explain how to run as a service on Linux

### Docker
You can also run this as a container by pulling the image from https://hub.docker.com/r/rokett/ldap-query.  You will need to deal with where to store the config file though.

**TODO** - Document config file stuff for Docker

## How to setup local Dev environment
This project maintains dependencies under version control, using Go modules, so building is really easy.

1. `go get github.com/rokett/ldap-query`.

You can now carry out development.

To build the executable, just run `make` from the root of the repository.

## Dockerfile
This Dockerfile will create a container that will set the entrypoint as `/ldap-queryd` so you can just pass in the command line options mentioned above to the container without needing to call the executable


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.2.2] - 2021/11/04
### Fixed
- #12 Returning all members a user is a group of, instead of just the first one.

## [1.2.1] - 2021/10/28
### Fixed
- Added missing import package to `main_linux.go`.

## [1.2.0] - 2021/10/28
### Added
- #9 Added CORS options to allow for lookups from websites.

## [1.1.0] - 2021/02/12
### Changed
- #5 Changed search to do a paged search in order to ensure we can return more than 1000 records in a single search.

## [1.0.1] - 2021/02/11
### Fixed
- Resolved an issue whereby the docker image was unable to be built because it was pointing to an old, pre Go module, release and was not copying the executable properly.

### Changed
- Added query details to logging if the query fails.

## [1.0.0] - 2020/05/08
Initial release


cmd/ldap-queryd/main_windows.go
package main

import (
	"flag"
	"fmt"
	"net"
	"net/http"
	"os"

	"github.com/Freman/eventloghook"
	"github.com/justinas/alice"
	"github.com/kardianos/service"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/rs/cors"
	"github.com/sirupsen/logrus"
	"golang.org/x/sys/windows/svc/eventlog"
)

type adQueryContextKeyType string

const clientIPCtxKey adQueryContextKeyType = "client_ip"
const traceIDCtxKey adQueryContextKeyType = "trace_id"

var (
	app           = "LDAP-Query"
	version       string
	build         string
	serviceDesc   = "REST API gateway for running queries against LDAP directory"
	directoryPort = 389

	versionFlg            = flag.Bool("version", false, "Display application version")
	winServiceCommand     = flag.String("service", "", "Manage Windows services: install, uninstall, start, stop")
	portFlg               = flag.Int("port", 9999, "Port to listen for requests on")
	debugFlg              = flag.Bool("debug", false, "Enable debug logging")
	allowedSourcesFlg     = flag.String("allowed_sources", "", "IPs for sources that need to be able to make queries")
	directoryHostsFlg     = flag.String("directory_hosts", "", "LDAP hosts to query")
	directoryBindDnFlg    = flag.String("directory_bind_dn", "", "DN of account used to bind to the directory")
	directoryBindPwdFlg   = flag.String("directory_bind_pw", "", "Password for account used to bind to the directory")
	corsAllowedOriginsFlg = flag.String("cors-allowed-origins", "", "Allowed origins for CORS purposes")
	corsAllowedHeadersFlg = flag.String("cors-allowed-headers", "*", "Allowed headers for CORS purposes")
	helpFlg               = flag.Bool("help", false, "Display application help")
)

type program struct {
	logger *logrus.Entry
}

//TODO Tracing

func main() {
	logrusLogger := logrus.New()

	logrusLogger.Out = os.Stdout
	logrusLogger.Formatter = &logrus.JSONFormatter{}

	logger := logrusLogger.WithFields(logrus.Fields{
		"version": version,
		"build":   build,
	})

	svcConfig := &service.Config{
		Name:        app,
		DisplayName: app,
		Description: serviceDesc,
	}

	prg := &program{
		logger: logger,
	}

	svc, err := service.New(prg, svcConfig)
	if err != nil {
		logger.WithFields(logrus.Fields{
			"function": "main",
			"error":    err,
		}).Fatal("error creating new service")
	}

	errs := make(chan error, 5)

	go func() {
		for {
			err := <-errs
			if err != nil {
				logger.WithFields(logrus.Fields{
					"function": "main",
					"error":    err,
				}).Fatal("unknown error")
			}
		}
	}()

	if *winServiceCommand != "" {
		err := service.Control(svc, *winServiceCommand)
		if err != nil {
			logger.WithFields(logrus.Fields{
				"function": "main",
				"error":    err,
				"control":  *winServiceCommand,
			}).Error("error controlling service")
		}

		os.Exit(0)
	} else {
		err = svc.Run()
		if err != nil {
			logger.WithFields(logrus.Fields{
				"function": "main",
				"error":    err,
			}).Error("error running application interactively")
		}
	}
}

func (p *program) Start(s service.Service) error {
	// Start should not block. Do the actual work async.
	go p.run(s)
	return nil
}

func (p *program) run(svc service.Service) {
	// If the application is NOT running interactively, then it is running as a service and we want to send logs to the Event Log.
	if !service.Interactive() {
		var el *eventlog.Log

		el, err := eventlog.Open(app)
		if err != nil {
			p.logger.WithFields(logrus.Fields{
				"function": "run",
				"error":    err,
			}).Error("unable to open Event Log")

			err := eventlog.InstallAsEventCreate(app, eventlog.Error|eventlog.Warning|eventlog.Info)
			if err != nil {
				p.logger.WithFields(logrus.Fields{
					"function": "run",
					"error":    err,
				}).Fatal("unable to create event source")
			}

			el, err = eventlog.Open(app)
			if err != nil {
				p.logger.WithFields(logrus.Fields{
					"function": "run",
					"error":    err,
				}).Fatal("unable to open event log after creating event source")
			}
		}
		defer el.Close()

		p.logger.Logger.Hooks.Add(eventloghook.NewHook(el))
	}

	flag.Parse()

	if *versionFlg {
		fmt.Printf("%s v%s build %s\n", app, version, build)
		os.Exit(0)
	}

	if *helpFlg {
		flag.PrintDefaults()
		os.Exit(0)
	}

	missingFlags := false

	if *allowedSourcesFlg == "" {
		missingFlags = true
		p.logger.WithFields(logrus.Fields{
			"function": "run",
		}).Error("The allowed_sources flag is required")
	}

	if *directoryHostsFlg == "" {
		missingFlags = true
		p.logger.WithFields(logrus.Fields{
			"function": "run",
		}).Error("The directory_hosts flag is required")
	}

	if *directoryBindDnFlg == "" {
		missingFlags = true
		p.logger.WithFields(logrus.Fields{
			"function": "run",
		}).Error("The directory_bind_dn flag is required")
	}

	if *directoryBindPwdFlg == "" {
		missingFlags = true
		p.logger.WithFields(logrus.Fields{
			"function": "run",
		}).Error("The directory_bind_pw flag is required")
	}

	if missingFlags {
		flag.PrintDefaults()
		os.Exit(1)
	}

	config := parseConfig(
		p.logger,
		*allowedSourcesFlg,
		*portFlg,
		*debugFlg,
		*directoryHostsFlg,
		*directoryBindDnFlg,
		*directoryBindPwdFlg,
		directoryPort,
		*corsAllowedOriginsFlg,
		*corsAllowedHeadersFlg,
	)

	if config.Server.Debug {
		p.logger.Logger.Level = logrus.DebugLevel
	} else {
		p.logger.Logger.Level = logrus.InfoLevel
	}

	// Need to ensure that we can bind to the directory before we bother listening for any requests.
	ldapConn, err := bindToDC(config.Directory, p.logger)
	if err != nil {
		p.logger.WithFields(logrus.Fields{
			"function": "run",
			"error":    err,
		}).Fatal("unable to bind to directory")
	}
	ldapConn.Close()

	listeningPort := fmt.Sprintf(":%d", config.Server.Port)
	server, err := net.Listen("tcp", listeningPort)
	if err != nil {
		p.logger.WithFields(logrus.Fields{
			"port":  listeningPort,
			"error": err,
		}).Fatal("unable to listen")
	}
	defer server.Close()

	middlewareChain := alice.New(
		checkMethodIsPOST, // Ensure method is allowed
		getClientIP,       // Store original client IP address in context
		checkRequestSource(config.Server.AllowedSources, p.logger), // Ensure source IP is allowed to query
		traceID(p.logger), // Generate Trace ID and store in context
	)

	// Using a locally scoped ServerMux to ensure that the only routes that can be registered are our own
	mux := http.NewServeMux()

	mux.Handle("/status", status())
	mux.Handle("/", middlewareChain.ThenFunc(search(config.Directory, p.logger)))
	mux.Handle("/metrics", promhttp.Handler())

	var handler http.Handler

	if len(config.Server.CorsAllowedOrigins) > 0 {
		opts := cors.Options{
			AllowedOrigins: config.Server.CorsAllowedOrigins,
			AllowedMethods: []string{"GET", "POST", "OPTIONS"},
			AllowedHeaders: config.Server.CorsAllowedHeaders,
		}

		co := cors.New(opts)

		handler = co.Handler(mux)
	} else {
		handler = http.Handler(mux)
	}

	p.logger.WithFields(logrus.Fields{
		"function": "run",
		"port":     listeningPort,
	}).Debug("API server listening")

	err = http.Serve(server, handler)
	if err != nil {
		p.logger.WithFields(logrus.Fields{
			"function": "run",
			"error":    err,
		}).Fatal("unable to start server")
	}
}

func (p *program) Stop(s service.Service) error {
	p.logger.Info("Stopping")
	// Stop should not block. Return with a few seconds.
	return nil
}
